### ISSUE TITLE
[Multipass] W/ FishyFacePunch - GetConnectionAddress() Error 

### ISSUE THREAD

**Issue by Katerlad** [FirstGearGames/FishNet#102]
When trying to get an Address from a FishyFacePunch Client using Multipass, it appears Multipass passes in the MultipassID instead of the transports clientID. 

This does not work specifically with a hostclient on  fishyfacepunch because the transportID is short.maxvalue, and the multipassID is 0. 

Upon further investigation it seems there may need to be some cleanup on Multipass with ClientIDs as there are a few areas where the logic seems flipped. 

```
ConnectionId 0 is invalid; address cannot be returned.
UnityEngine.Debug:LogError (object)
FishyFacepunch.Server.ServerSocket:GetConnectionAddress (int) (at Assets/Plugins/Core/ServerSocket.cs:353)
FishyFacepunch.FishyFacepunch:GetConnectionAddress (int) (at Assets/Plugins/FishyFacepunch.cs:145)
FishNet.Transporting.Multipass.Multipass:GetConnectionAddress (int) (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:238)
FishNet.Connection.NetworkConnection:GetAddress () (at Assets/FishNet/Runtime/Connection/NetworkConnection.QOL.cs:24)
GutzClientManager:HandleRemoteClientConnection (FishNet.Connection.NetworkConnection,FishNet.Transporting.RemoteConnectionStateArgs) (at Assets/Scripts/Networking/GutzClientManager.cs:55)
```

This gets called from the NetworkConnection.GetConnectionAddress which calls Multipass.GetConnectionAddress(connectionid) 

```
public override string GetConnectionAddress(int connectionId)
        {
            TransportIdData data;
            if (!GetTransportIdData(connectionId, out data))
                return string.Empty;

            return _transports[data.TransportIndex].GetConnectionAddress(connectionId);
        } 
```

``` private void Multipass_OnRemoteConnectionState(RemoteConnectionStateArgs connectionStateArgs)
        {
            /* When starting Multipass needs to get a new
            * connectionId to be used within FN. This is the 'ClientId'
             * that is passed around for ownership, rpcs, ect.
             * 
             * The new connectionId will be linked with the connectionId
             * from the transport, named transportConnectionid. 
             * 
             * When data arrives the transportStateId is used as a key
             * in fromClientIds, where Multipass Id is returned. The argument values
             * are then overwritten with the MultipassId.
             * 
             * When data is being sent the same process is performed but reversed.
             * The connectionId is looked up in toClientIds, where the transportConnectionId
             * is output. Then as before the argument values are overwritten with the
             * transportConnectionId. */



            int transportIndex = connectionStateArgs.TransportIndex;
            int transportId = connectionStateArgs.ConnectionId;
            int multipassId;
            Dictionary<int, int> transportToMultipassDict = _transportToMultipass[transportIndex];
  //Started.
            if (connectionStateArgs.ConnectionState == RemoteConnectionState.Started)
            {
                multipassId = _availableIds.Dequeue();
                transportToMultipassDict[transportId] = multipassId;
                _multipassToTransport[multipassId] = new TransportIdData(transportId, transportIndex);
            }
            //Stopped.
            else
            {
                if (!GetMultipassId(transportIndex, transportId, out multipassId))
                    return;

                _availableIds.Enqueue(multipassId);
                transportToMultipassDict.Remove(transportId);
                _multipassToTransport.Remove(multipassId);
            }

            connectionStateArgs.ConnectionId = multipassId;
            OnRemoteConnectionState?.Invoke(connectionStateArgs);
        }
```

One of the areas that I believe has the logic incorrect is the transportIdData struct. This should hold a Transport Client ID and not a MultipassID. Which leaves me to believe that this class should be investigated for other areas this may bleed into problems.

```public struct TransportIdData
        {
            public int MultipassId;
            public int TransportIndex;

            public TransportIdData(int multipassId, int transportIndex)
            {
                MultipassId = multipassId;
                TransportIndex = transportIndex;
            }
        } 
```

**Comment by lordubbe** [comment]
Stumbled across this issue by chance, and just thought I'd second the need to have a fresh look into multipass.
I too have experienced errors with multipass close to this. Can't post it right now as it's been a while since I was on the multiplayer branch of my game and I am currently "on the way there" but fixing huge merge conflicts. 
I will make sure to update when I can post the error!

**Comment by lordubbe** [comment]
Here we go, the error is `Multipass connectionId could not be found for transportIndex 0, connectionId of 32767.`
Stack trace:
`FishNet.Transporting.Multipass.Multipass:GetMultipassId (int,int,int&) (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:200)
FishNet.Transporting.Multipass.Multipass:Multipass_OnRemoteConnectionState (FishNet.Transporting.RemoteConnectionStateArgs) (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:364)
FishyFacepunch.FishyFacepunch:HandleRemoteConnectionState (FishNet.Transporting.RemoteConnectionStateArgs) (at Assets/FishNet/Plugins/FishyFacepunch/FishyFacepunch.cs:200)
FishyFacepunch.Server.ServerSocket:OnClientHostState (bool) (at Assets/FishNet/Plugins/FishyFacepunch/Core/ServerSocket.cs:396)
FishyFacepunch.Client.ClientHostSocket:SetLocalConnectionState (FishNet.Transporting.LocalConnectionState,bool) (at Assets/FishNet/Plugins/FishyFacepunch/Core/ClientHostSocket.cs:66)
FishyFacepunch.Client.ClientHostSocket:StartConnection (FishyFacepunch.Server.ServerSocket) (at Assets/FishNet/Plugins/FishyFacepunch/Core/ClientHostSocket.cs:53)
FishyFacepunch.FishyFacepunch:StartClient (string) (at Assets/FishNet/Plugins/FishyFacepunch/FishyFacepunch.cs:464)
FishyFacepunch.FishyFacepunch:StartConnection (bool) (at Assets/FishNet/Plugins/FishyFacepunch/FishyFacepunch.cs:345)
FishNet.Transporting.Multipass.Multipass:StartClient () (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:881)
FishNet.Transporting.Multipass.Multipass:StartConnection (bool,int) (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:751)
FishNet.Transporting.Multipass.Multipass:StartConnection (bool) (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:732)
FishNet.Managing.Client.ClientManager:StartConnection () (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:177)
Golfie.Multiplayer.PlatformNetworkManagerBase/<StartClient>d__16:MoveNext () (at Assets/PlatformNetworkManagerBase.cs:71)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:Start<Golfie.Multiplayer.PlatformNetworkManagerBase/<StartClient>d__16> (Golfie.Multiplayer.PlatformNetworkManagerBase/<StartClient>d__16&)
Golfie.Multiplayer.PlatformNetworkManagerBase:StartClient (LobbyData)
Golfie.Multiplayer.PlatformNetworkManagerBase/<JoinLobby>d__13:MoveNext () (at Assets/PlatformNetworkManagerBase.cs:47)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Golfie.Multiplayer.LobbyJoinResult>:SetResult (Golfie.Multiplayer.LobbyJoinResult)
Golfie.Multiplayer.SteamNetworkManager/<JoinLobbyInternal>d__17:MoveNext () (at Assets/SteamNetworkManager.cs:112)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<System.Nullable`1<Steamworks.Data.Lobby>>:SetResult (System.Nullable`1<Steamworks.Data.Lobby>)
Steamworks.SteamMatchmaking/<JoinLobbyAsync>d__46:MoveNext () (at D:/a/Facepunch.Steamworks/Facepunch.Steamworks/Facepunch.Steamworks/SteamMatchmaking.cs:170)
System.Runtime.CompilerServices.AsyncMethodBuilderCore/MoveNextRunner:Run ()
Steamworks.Dispatch:ProcessResult (Steamworks.Dispatch/CallbackMsg_t) (at D:/a/Facepunch.Steamworks/Facepunch.Steamworks/Facepunch.Steamworks/Classes/Dispatch.cs:228)
Steamworks.Dispatch:ProcessCallback (Steamworks.Dispatch/CallbackMsg_t,bool) (at D:/a/Facepunch.Steamworks/Facepunch.Steamworks/Facepunch.Steamworks/Classes/Dispatch.cs:138)
Steamworks.Dispatch:Frame (Steamworks.Data.HSteamPipe) (at D:/a/Facepunch.Steamworks/Facepunch.Steamworks/Facepunch.Steamworks/Classes/Dispatch.cs:102)
Steamworks.Dispatch/<LoopClientAsync>d__22:MoveNext () (at D:/a/Facepunch.Steamworks/Facepunch.Steamworks/Facepunch.Steamworks/Classes/Dispatch.cs:240)
UnityEngine.UnitySynchronizationContext:ExecuteTasks ()`

**Comment by Katerlad** [comment]
What version of fishy face punch are you using?

**Comment by FirstGearGames** [comment]
I'll see if they can close down that git to prevent future confusion. But yes, please try to replicate on 2.1 and let me know if you're still having `Multipass connectionId could not be found for transportIndex 0, connectionId of 32767.` after.

Per our talk Katerlad, I'm aware you are experiencing your different problem on 2.1.

**Comment by Katerlad** [comment]
Yes I am on 2.1.0 all details from our convo are above for convienience, it doesn't seem like the issues found in multipass are affecting anything else, but I do worry we may stumble across a few areas where the client ID logic needs to be cleaned up or it will affect other things. 

Hopefully it's just getconnectionaddress() and that transportIdStruct having multipassID instead of TransportID.

**Comment by lordubbe** [comment]
I've deleted my comments for good clarity. After upgrading both FN and FishyFacepunch I no longer get the issue I described and see that it was recently fixed :) Just leaving this comment for the sake of documentation.

**Comment by FirstGearGames** [comment]
> I've deleted my comments for good clarity. After upgrading both FN and FishyFacepunch I no longer get the issue I described and see that it was recently fixed :) Just leaving this comment for the sake of documentation.

Okay thanks. I'm keeping this open though because Katerlad is experiencing something else.

**Comment by FirstGearGames** [comment]
> When trying to get an Address from a FishyFacePunch Client using Multipass, it appears Multipass passes in the MultipassID instead of the transports clientID.
> 
> This does not work specifically with a hostclient on fishyfacepunch because the transportID is short.maxvalue, and the multipassID is 0.
> 
> Upon further investigation it seems there may need to be some cleanup on Multipass with ClientIDs as there are a few areas where the logic seems flipped.
> 
> ```
> ConnectionId 0 is invalid; address cannot be returned.
> UnityEngine.Debug:LogError (object)
> FishyFacepunch.Server.ServerSocket:GetConnectionAddress (int) (at Assets/Plugins/Core/ServerSocket.cs:353)
> FishyFacepunch.FishyFacepunch:GetConnectionAddress (int) (at Assets/Plugins/FishyFacepunch.cs:145)
> FishNet.Transporting.Multipass.Multipass:GetConnectionAddress (int) (at Assets/FishNet/Runtime/Transporting/Transports/Multipass/Multipass.cs:238)
> FishNet.Connection.NetworkConnection:GetAddress () (at Assets/FishNet/Runtime/Connection/NetworkConnection.QOL.cs:24)
> GutzClientManager:HandleRemoteClientConnection (FishNet.Connection.NetworkConnection,FishNet.Transporting.RemoteConnectionStateArgs) (at Assets/Scripts/Networking/GutzClientManager.cs:55)
> ```
> 
> This gets called from the NetworkConnection.GetConnectionAddress which calls Multipass.GetConnectionAddress(connectionid)
> 
> ```
> public override string GetConnectionAddress(int connectionId)
>         {
>             TransportIdData data;
>             if (!GetTransportIdData(connectionId, out data))
>                 return string.Empty;
> 
>             return _transports[data.TransportIndex].GetConnectionAddress(connectionId);
>         } 
> ```
> 
> ```
>         {
>             /* When starting Multipass needs to get a new
>             * connectionId to be used within FN. This is the 'ClientId'
>              * that is passed around for ownership, rpcs, ect.
>              * 
>              * The new connectionId will be linked with the connectionId
>              * from the transport, named transportConnectionid. 
>              * 
>              * When data arrives the transportStateId is used as a key
>              * in fromClientIds, where Multipass Id is returned. The argument values
>              * are then overwritten with the MultipassId.
>              * 
>              * When data is being sent the same process is performed but reversed.
>              * The connectionId is looked up in toClientIds, where the transportConnectionId
>              * is output. Then as before the argument values are overwritten with the
>              * transportConnectionId. */
> 
>             int transportIndex = connectionStateArgs.TransportIndex;
>             int transportId = connectionStateArgs.ConnectionId;
>             int multipassId;
>             Dictionary<int, int> transportToMultipassDict = _transportToMultipass[transportIndex];
>   //Started.
>             if (connectionStateArgs.ConnectionState == RemoteConnectionState.Started)
>             {
>                 multipassId = _availableIds.Dequeue();
>                 transportToMultipassDict[transportId] = multipassId;
>                 _multipassToTransport[multipassId] = new TransportIdData(transportId, transportIndex);
>             }
>             //Stopped.
>             else
>             {
>                 if (!GetMultipassId(transportIndex, transportId, out multipassId))
>                     return;
> 
>                 _availableIds.Enqueue(multipassId);
>                 transportToMultipassDict.Remove(transportId);
>                 _multipassToTransport.Remove(multipassId);
>             }
> 
>             connectionStateArgs.ConnectionId = multipassId;
>             OnRemoteConnectionState?.Invoke(connectionStateArgs);
>         }
> ```
> 
> One of the areas that I believe has the logic incorrect is the transportIdData struct. This should hold a Transport Client ID and not a MultipassID. Which leaves me to believe that this class should be investigated for other areas this may bleed into problems.
> 
> ```
>         {
>             public int MultipassId;
>             public int TransportIndex;
> 
>             public TransportIdData(int multipassId, int transportIndex)
>             {
>                 MultipassId = multipassId;
>                 TransportIndex = transportIndex;
>             }
>         } 
> ```

I don't see it in the issue, and I know we talked about this previously but having the information here would be of great help, can you please provide me with reproduction steps. Thank you.

**Comment by Katerlad** [comment]
[Multipass Test.zip](https://github.com/FirstGearGames/FishNet/files/9628623/Multipass.Test.zip)

I attached a base test project above. 

Must have fishyfacepunch, and steamworks installed. 
I wired up the ServerButton to start the client and server using this:
`mp.StartConnection(true);
mp.StartConnection(false);`

I Then call a Debug in the GetConnectionAddress Script I added to print on the server the client ID that connected and what address its from. 

So just open up the project and click server and look at the second error. First error is just letting you know you should set what transport to use for client connection but it defaulted to fishyfacepunch.

**Comment by FirstGearGames** [comment]
The error is in this method 
```
        public override string GetConnectionAddress(int connectionId)
        {
            TransportIdData data;
            if (!GetTransportIdData(connectionId, out data))
                return string.Empty;

            return _transports[data.TransportIndex].GetConnectionAddress(connectionId);
        }```
        It should be using data.TransportId, (which was also named incorrectly to MultipassId), instead of connectionId.
        I resolved this and I'm double checking for other mistakes.

**Comment by Katerlad** [comment]
Perfect, that matches what I had mentioned in OG post! I was worried maybe other areas were affected by the struct being mis labeled with multi-pass ID instead of transport ID but I didn't look too hard myself.

**Comment by FirstGearGames** [comment]
Solved in 2.5.0
